indico/modules/auth/forms.py

```diff
@@ -46,22 +48,39 @@ def _check_not_blacklisted(form, field):
             raise ValidationError(error_msg)
 
 
+def _check_use_own_email_as_username(form, field):
+    if validate_email(field.data, check_dns=False):
+        emails = (UserEmail.query
+                  .filter(~User.is_pending,
+                          ~UserEmail.is_user_deleted,
+                          UserEmail.user == form.user,
+                          UserEmail.email == field.data)
+                  .join(User))
+        if not emails.has_rows():
+            raise ValidationError(_('Only emails addresses linked to your account can be used as a username.'))
+
+
 class LocalLoginForm(IndicoForm):
     identifier = StringField(_('Username'), [DataRequired()], filters=[_tolower])
     password = PasswordField(_('Password'), [DataRequired()])
 
 
 class AddLocalIdentityForm(IndicoForm):
-    username = StringField(_('Username'), [DataRequired(), _check_existing_username], filters=[_tolower])
+    username = StringField(_('Username'), [DataRequired(), _check_existing_username, _check_use_own_email_as_username],
+                           filters=[_tolower])
     password = PasswordField(_('Password'), [DataRequired(), SecurePassword('set-user-password',
                                                                             username_field='username')],
                              render_kw={'autocomplete': 'new-password'})
     confirm_password = PasswordField(_('Confirm password'), [DataRequired(), ConfirmPassword('password')],
                                      render_kw={'autocomplete': 'new-password'})
 
+    def __init__(self, *args, **kwargs):
+        self.user = kwargs.pop('user', None)
```

Comment:
```suggestion
    def __init__(self, *args, user=None, **kwargs):
        self.user = user
```

the kwargs pop is a leftover from python 2 times where you could not mix specific keyword(-only) args and `**kwargs`